File: crates/graphviz-anywhere/packages/react-native/android/src/main/java/com/graphviznative/GraphvizModule.java
/*
 * GraphvizModule.java
 *
 * React Native native module for Android.
 * Manages a singleton Graphviz context and dispatches rendering
 * to a background thread to keep the JS thread responsive.
 *
 * Licensed under the Apache License, Version 2.0
 */

package com.graphviznative;

import android.util.Base64;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GraphvizModule extends ReactContextBaseJavaModule {

    private static final String TAG = "GraphvizNative";
    private static final String MODULE_NAME = "GraphvizNative";

    private static final Set<String> TEXT_FORMATS = new HashSet<>(
        Arrays.asList("svg", "json", "dot", "xdot", "plain")
    );

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private long contextPtr = 0;

    static {
        System.loadLibrary("graphviz_jni");
    }

    public GraphvizModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    @NonNull
    public String getName() {
        return MODULE_NAME;
    }

    @Override
    public void invalidate() {
        if (contextPtr != 0) {
            nativeContextFree(contextPtr);
            contextPtr = 0;
        }
        executor.shutdown();
        super.invalidate();
    }

    /**
     * Lazily create the native context. Must be called from the executor thread.
     */
    private synchronized long ensureContext() {
        if (contextPtr == 0) {
            contextPtr = nativeContextNew();
            if (contextPtr == 0) {
                Log.e(TAG, "Failed to create Graphviz context");
            }
        }
        return contextPtr;
    }

    @ReactMethod
    public void renderDot(String dot, String engine, String format, Promise promise) {
        executor.execute(() -> {
            try {
                long ctx = ensureContext();
                if (ctx == 0) {
                    promise.reject("NOT_INITIALIZED", "Failed to initialize Graphviz context");
                    return;
                }

                String[] result = new String[1];
                int err = nativeRender(ctx, dot, engine, format, result);

                if (err != 0) {
                    String code = errorCodeToString(err);
                    String message = nativeStrerror(err);
                    promise.reject(code, message);
                    return;
                }

                if (result[0] == null) {
                    promise.reject("RENDER_FAILED", "Render returned null output");
                    return;
                }

                // For text formats, return as-is; for binary formats, already base64 from JNI
                promise.resolve(result[0]);

            } catch (Exception e) {
                Log.e(TAG, "Render failed with exception", e);
                promise.reject("UNKNOWN", "Unexpected error: " + e.getMessage());
            }
        });
    }

    @ReactMethod
    public void getVersion(Promise promise) {
        try {
            String version = nativeVersion();
            promise.resolve(version);
        } catch (Exception e) {
            promise.reject("UNKNOWN", "Failed to get version: " + e.getMessage());
        }
    }

    private static String errorCodeToString(int err) {
        switch (err) {
            case -1: return "NULL_INPUT";
            case -2: return "INVALID_DOT";
            case -3: return "LAYOUT_FAILED";
            case -4: return "RENDER_FAILED";
            case -5: return "INVALID_ENGINE";
            case -6: return "INVALID_FORMAT";
            case -7: return "OUT_OF_MEMORY";
            case -8: return "NOT_INITIALIZED";
            default: return "UNKNOWN";
        }
    }

    /*
     * JNI native methods.
     * These are implemented in the CMake-built native library that
     * wraps graphviz_api.h via JNI.
     */
    private static native long nativeContextNew();
    private static native void nativeContextFree(long ctx);
    private static native int nativeRender(long ctx, String dot, String engine,
                                            String format, String[] outResult);
    private static native String nativeStrerror(int err);
    private static native String nativeVersion();
}

File: crates/graphviz-anywhere/packages/react-native/ios/GraphvizModule.m
/*
 * GraphvizModule.m
 *
 * React Native native module implementation for iOS / macOS.
 * Manages a singleton Graphviz context and dispatches rendering
 * to a background queue to keep the JS thread responsive.
 *
 * Licensed under the Apache License, Version 2.0
 */

#import "GraphvizModule.h"
#import "graphviz_api.h"

#import <React/RCTLog.h>

@implementation GraphvizModule {
    gv_context_t *_context;
    dispatch_queue_t _renderQueue;
}

RCT_EXPORT_MODULE(GraphvizNative)

- (instancetype)init {
    self = [super init];
    if (self) {
        _renderQueue = dispatch_queue_create("com.graphviznative.render", DISPATCH_QUEUE_SERIAL);
        _context = NULL;
    }
    return self;
}

- (void)dealloc {
    if (_context) {
        gv_context_free(_context);
        _context = NULL;
    }
}

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

/**
 * Lazily initializes the Graphviz context on the render queue.
 * Must be called from within a dispatch to _renderQueue.
 */
- (gv_context_t *)ensureContext {
    if (!_context) {
        _context = gv_context_new();
        if (!_context) {
            RCTLogError(@"GraphvizNative: failed to create Graphviz context");
        }
    }
    return _context;
}

/**
 * Map native error codes to JS-friendly error code strings.
 */
static NSString *errorCodeToString(gv_error_t err) {
    switch (err) {
        case GV_ERR_NULL_INPUT:       return @"NULL_INPUT";
        case GV_ERR_INVALID_DOT:      return @"INVALID_DOT";
        case GV_ERR_LAYOUT_FAILED:    return @"LAYOUT_FAILED";
        case GV_ERR_RENDER_FAILED:    return @"RENDER_FAILED";
        case GV_ERR_INVALID_ENGINE:   return @"INVALID_ENGINE";
        case GV_ERR_INVALID_FORMAT:   return @"INVALID_FORMAT";
        case GV_ERR_OUT_OF_MEMORY:    return @"OUT_OF_MEMORY";
        case GV_ERR_NOT_INITIALIZED:  return @"NOT_INITIALIZED";
        default:                      return @"UNKNOWN";
    }
}

/**
 * Returns YES for text-based output formats that do not need base64 encoding.
 */
static BOOL isTextFormat(NSString *format) {
    static NSSet *textFormats;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        textFormats = [NSSet setWithArray:@[@"svg", @"json", @"dot", @"xdot", @"plain"]];
    });
    return [textFormats containsObject:format];
}

RCT_EXPORT_METHOD(renderDot:(NSString *)dot
                  engine:(NSString *)engine
                  format:(NSString *)format
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    dispatch_async(_renderQueue, ^{
        gv_context_t *ctx = [self ensureContext];
        if (!ctx) {
            reject(@"NOT_INITIALIZED", @"Failed to initialize Graphviz context", nil);
            return;
        }

        char *outData = NULL;
        size_t outLength = 0;

        gv_error_t err = gv_render(
            ctx,
            [dot UTF8String],
            [engine UTF8String],
            [format UTF8String],
            &outData,
            &outLength
        );

        if (err != GV_OK) {
            NSString *code = errorCodeToString(err);
            NSString *message = [NSString stringWithUTF8String:gv_strerror(err)];
            reject(code, message, nil);
            if (outData) {
                gv_free_render_data(outData);
            }
            return;
        }

        NSString *result;
        if (isTextFormat(format)) {
            result = [[NSString alloc] initWithBytes:outData
                                              length:outLength
                                            encoding:NSUTF8StringEncoding];
        } else {
            NSData *data = [NSData dataWithBytesNoCopy:outData
                                                length:outLength
                                          freeWhenDone:NO];
            result = [data base64EncodedStringWithOptions:0];
        }

        gv_free_render_data(outData);

        if (!result) {
            reject(@"RENDER_FAILED", @"Failed to convert render output to string", nil);
            return;
        }

        resolve(result);
    });
}

RCT_EXPORT_METHOD(getVersion:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    const char *version = gv_version();
    if (version) {
        resolve([NSString stringWithUTF8String:version]);
    } else {
        reject(@"UNKNOWN", @"Failed to get Graphviz version", nil);
    }
}

#ifdef RCT_NEW_ARCH_ENABLED
- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params {
    return std::make_shared<facebook::react::NativeGraphvizNativeSpecJSI>(params);
}
#endif

@end
